<a href="https://colab.research.google.com/github/Ikunalk4/ml/blob/main/neenopal_task_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### There are 2 datasets present in the file. Data 1 and Data 2. The primary key for both data1 and data2 is Order Id + Product ID combination (i.e. the individual datasets do not have any duplicate on this combination)


Let us import the libraries, We will use pandas.

In [1]:
import pandas as pd

Let us first upload the data file which is required to analyse:-  

In [2]:
df1 = pd.read_csv('/content/Records Matching Task.xlsx - data1.csv')
df2 = pd.read_csv('/content/Records Matching Task.xlsx - data2.csv')

#### 1) How to identify the Records (Order ID + Product ID combination) present in data1 but missing in data2 (Specify the number of records missing in your answer)

In [3]:
df1.head()

Order ID       Product ID  Qty
0  CA-2014-100006  TEC-PH-10002075    3
1  CA-2014-100090  FUR-TA-10003715    3
2  CA-2014-100090  OFF-BI-10001597    6
3  CA-2014-100293  OFF-PA-10000176    6
4  CA-2014-100328  OFF-BI-10000343    1

In [4]:
df2.head()

Order ID       Product ID  Qty
0  CA-2014-100090  FUR-TA-10003715    3
1  CA-2014-100090  OFF-BI-10001597    6
2  CA-2014-100293  OFF-PA-10000176    6
3  CA-2014-100328  OFF-BI-10000343    1
4  CA-2014-100363  OFF-FA-10000611    2

In [5]:
df1.shape

(9478, 3)

In [6]:
df2.shape

(9479, 3)

In [7]:
A = pd.merge(df1, df2, how="outer", indicator=True)
A

Order ID       Product ID  Qty      _merge
0     CA-2014-100006  TEC-PH-10002075    3   left_only
1     CA-2014-100090  FUR-TA-10003715    3        both
2     CA-2014-100090  OFF-BI-10001597    6        both
3     CA-2014-100293  OFF-PA-10000176    6        both
4     CA-2014-100328  OFF-BI-10000343    1        both
...              ...              ...  ...         ...
9981  US-2017-165344  OFF-BI-10003196   10  right_only
9982  US-2017-165358  TEC-CO-10001943    5  right_only
9983  US-2017-167920  OFF-AP-10000159    5  right_only
9984  US-2017-169502  OFF-AP-10001947    5  right_only
9985  US-2017-169551  FUR-BO-10001519    3  right_only

[9986 rows x 4 columns]

In [9]:
Left_only = A[A['_merge'] == 'left_only']
Left_only

Order ID       Product ID  Qty     _merge
0     CA-2014-100006  TEC-PH-10002075    3  left_only
10    CA-2014-100678  OFF-EN-10000056    3  left_only
19    CA-2014-100895  OFF-AR-10004511    2  left_only
35    CA-2014-101560  OFF-BI-10000309    3  left_only
61    CA-2014-102673  OFF-LA-10001771   12  left_only
...              ...              ...  ...        ...
9390  US-2017-160836  OFF-AP-10001626    2  left_only
9403  US-2017-162558  FUR-FU-10002364    2  left_only
9420  US-2017-163657  OFF-BI-10000138    5  left_only
9427  US-2017-164056  FUR-TA-10001307    5  left_only
9435  US-2017-165456  FUR-CH-10003981    6  left_only

[507 rows x 4 columns]

#### The number of records present in data1 but missing in data2 is 508.

In [10]:
Left_only.shape

(507, 4)

#### 2) How to identify the Records (Order ID + Product ID combination) missing in data1 but present in data2 (Specify the number of records missing in your answer)

In [11]:
Right_only = A[A['_merge'] == 'right_only']
Right_only

Order ID       Product ID  Qty      _merge
9478  CA-2014-100706  TEC-AC-10001314    2  right_only
9479  CA-2014-100762  OFF-PA-10001815    3  right_only
9480  CA-2014-101427  OFF-AR-10002257    3  right_only
9481  CA-2014-102652  FUR-FU-10001918    7  right_only
9482  CA-2014-102869  OFF-PA-10000788    3  right_only
...              ...              ...  ...         ...
9981  US-2017-165344  OFF-BI-10003196   10  right_only
9982  US-2017-165358  TEC-CO-10001943    5  right_only
9983  US-2017-167920  OFF-AP-10000159    5  right_only
9984  US-2017-169502  OFF-AP-10001947    5  right_only
9985  US-2017-169551  FUR-BO-10001519    3  right_only

[508 rows x 4 columns]

#### The number of records missing in data2 but present in data1 is 507.

In [12]:
Right_only.shape

(508, 4)

#### 3) Find the Sum of the total Qty of Records missing in data1 but present in data2.

In [13]:
Sum_of_the_total_Qty = Right_only.Qty.sum()
Sum_of_the_total_Qty

1956

#### The sum of total Qty is 1956.

#### 4) Find the total number of unique records (Order ID + Product ID combination) present in the combined dataset of data1 and data2.

In [14]:
Unique_value = A
Unique_value

Order ID       Product ID  Qty      _merge
0     CA-2014-100006  TEC-PH-10002075    3   left_only
1     CA-2014-100090  FUR-TA-10003715    3        both
2     CA-2014-100090  OFF-BI-10001597    6        both
3     CA-2014-100293  OFF-PA-10000176    6        both
4     CA-2014-100328  OFF-BI-10000343    1        both
...              ...              ...  ...         ...
9981  US-2017-165344  OFF-BI-10003196   10  right_only
9982  US-2017-165358  TEC-CO-10001943    5  right_only
9983  US-2017-167920  OFF-AP-10000159    5  right_only
9984  US-2017-169502  OFF-AP-10001947    5  right_only
9985  US-2017-169551  FUR-BO-10001519    3  right_only

[9986 rows x 4 columns]

In [15]:
Unique_value.head()

Order ID       Product ID  Qty     _merge
0  CA-2014-100006  TEC-PH-10002075    3  left_only
1  CA-2014-100090  FUR-TA-10003715    3       both
2  CA-2014-100090  OFF-BI-10001597    6       both
3  CA-2014-100293  OFF-PA-10000176    6       both
4  CA-2014-100328  OFF-BI-10000343    1       both

In [16]:
Unique_value.shape

(9986, 4)

Total number of unique values are 9986.